File: src/main/java/org/kite9/diagram/common/BiDirectional.java
package org.kite9.diagram.common;

import org.kite9.diagram.position.Direction;

/**
 * This generic interface allows you to specify a bi-directional, optionally directed from-to relationships, where from and to are 
 * both objects of generic class X.
 * 
 * @author robmoffat
 *
 * @param <X>
 */
public interface BiDirectional<X> {

	public X getFrom();

	public X getTo();

	public void setFrom(X v);

	public void setTo(X v);

	/**
	 * Returns from, if to is the argument, or to if from is the argument.
	 * @param end
	 * @return
	 */
	public X otherEnd(X end);

	public boolean meets(BiDirectional<X> e);

	public boolean meets(X v);
	
	/**
	 * Indicates the layout of from/to for the bi-directional item.  If this is non-null, then it is describing
	 * the single direction it needs to flow in for the diagram.
	 */
	public Direction getDrawDirection();
	
	public Direction getDrawDirectionFrom(X from);
	
	public void setDrawDirection(Direction d);
	
	public void setDrawDirectionFrom(Direction d, X from);


}

File: src/main/kotlin/org/kite9/diagram/common/BiDirectional.kt
package org.kite9.diagram.common

import org.kite9.diagram.model.position.Direction

/**
 * This generic interface allows you to specify a bi-directional, optionally directed from-to relationships, where from and to are
 * both objects of generic class X.
 *
 * @author robmoffat
 *
 * @param <X>
</X> */
interface BiDirectional<X> {

    fun getFrom(): X
    fun getTo(): X

    /**
     * Returns from, if to is the argument, or to if from is the argument.
     * @param end
     * @return
     */
    fun otherEnd(end: X): X
    fun meets(e: BiDirectional<X>?): Boolean
    fun meets(v: X): Boolean

    /**
     * Indicates the layout of from/to for the bi-directional item.  If this is non-null, then it is describing
     * the single direction it needs to flow in for the diagram.
     */
    fun getDrawDirection(): Direction?
    fun getDrawDirectionFrom(from: X): Direction?

    /**
     * Every BiDirectional should have a unique ID.
     */
    fun getID(): String

}